cscs-checks/apps/quantumespresso/quantumespresso_check.py

```diff
@@ -5,51 +5,133 @@
+    },
+    'pilatus:mc': {
+        'maint': (53.0, None, 0.10, 's'),
+        'prod': (53.0, None, 0.10, 's')
+    },
+}
+
+REFERENCE_GPU_PERFORMANCE_SMALL = {
+    'dom:mc': {
+        'maint': (61.0, None, 0.05, 's'),
+        'prod': (61.0, None, 0.05, 's')
+    },
+    'daint:mc': {
+        'maint': (61.0, None, 0.05, 's'),
+        'prod': (61.0, None, 0.05, 's')
+    }
+}
+
+REFERENCE_GPU_PERFORMANCE_LARGE = {
+    'daint:mc': {
+        'maint': (54.0, None, 0.05, 's'),
+        'prod': (54.0, None, 0.05, 's')
+    }
+}
+
+
+class QuantumESPRESSOCheck(QuantumESPRESSO):
     scale = parameter(['small', 'large'])
-    variant = parameter(['maint', 'prod'])
+    benchmark = parameter(['maint', 'prod'])
+
+    modules = ['QuantumESPRESSO']
+    executable = 'pw.x'
+    input_file = 'ausurf.in'
+    maintainers = ['LM']
+    tags = {'scs'}
+    strict_check = False
+    extra_resources = {
+        'switches': {
+            'num_switches': 1
+        }
+    }
 
-    def __init__(self):
+    @run_after('init')
+    def env_define(self):
         if self.current_system.name in ['eiger', 'pilatus']:
             self.valid_prog_environs = ['cpeIntel']
         else:
             self.valid_prog_environs = ['builtin']
 
-        self.modules = ['QuantumESPRESSO']
-        self.executable = 'pw.x'
-        self.executable_opts = ['-in', 'ausurf.in']
-
-        self.sanity_patterns = sn.all([
-            sn.assert_found(r'convergence has been achieved', self.stdout),
-        ])
+    @run_after('init')
+    def set_reference(self):
+        self.reference = references[self.benchmark]
 
-        self.perf_patterns = {
-            'time': sn.extractsingle(r'electrons.+\s(?P<wtime>\S+)s WALL',
-                                     self.stdout, 'wtime', float)
+    @run_after('init')
+    def set_tags(self):
+        self.tags |= {
+            'maintenance' if self.benchmark == 'maint' else 'production'
         }
 
-        self.maintainers = ['LM']
-        self.tags = {'scs'}
-        self.strict_check = False
-        self.extra_resources = {
-            'switches': {
-                'num_switches': 1
-            }
-        }
+    @run_after('setup')
+    def set_generic_perf_references(self):
+        self.reference.update({'*': {
+            self.benchmark: (0, None, None, 's')
+        }})
 
+    @run_after('setup')
+    def set_perf_patterns(self):
+        self.perf_patterns = {
+            self.benchmark: sn.extractsingle(
+                r'electrons.+\s(?P<wtime>\S+)s WALL',
+                self.stdout, 'wtime', float)
+        }
 
 @rfm.simple_test
 class QuantumESPRESSOCpuCheck(QuantumESPRESSOCheck):
-    def __init__(self):
-        super().__init__()
+    valid_systems = ['daint:mc', 'eiger:mc', 'pilatus:mc']
+
+    @run_after('init')
+    def set_reference(self):
+        if self.scale == 'small':
+            self.reference = REFERENCE_CPU_PERFORMANCE_SMALL
+            self.energy_value = -11427.09017168
```

Comment:
This information is input-file dependent. Therefore, I think that this should move together to the base class.